### ISSUE TITLE
tmp issue

### ISSUE THREAD

**Issue by doublejy715** [doublejy715/review_StyleMapGAN#1]
근데 이거 원래 하던 parrallel과 다른 점은?

# Swaping subprocessing
[홈페이지](https://pytorch.org/docs/stable/multiprocessing.html)

multi gpu 에서 학습시키기 위한 코드.
function을 여러 gpu에 올려서 학습시키기 위해서 사용한다.

torch.multiprocessing.spawn

- fn : spawned process할 function을 의미. 
fn(i,*args)
i : process index
args : 넣어줄 매개변수
- args : fn에 넣어줄 매개변수. tuple로 묶어서 넣어준다.
- nprocs : 몇개의 sub process를 만들 것인지(보통 gpus의 개수로 설정)
- join : Perform a blocking join on all processes.(blocking join 이란?)

# dist.init_process_group

[공식 홈페이지 튜토리얼](https://tutorials.pytorch.kr/intermediate/dist_tuto.html)

# torch.utils.data.distributed.DistributedSampler
[ torch 공식 설명 ](https://pytorch.org/docs/stable/data.html)
[ 부가 설명 ](https://subinium.github.io/pytorch-dataloader/)
[ 부가 설명 2 ](https://medium.com/daangn/pytorch-multi-gpu-%ED%95%99%EC%8A%B5-%EC%A0%9C%EB%8C%80%EB%A1%9C-%ED%95%98%EA%B8%B0-27270617936b)

병렬 학습할때(분산 처리시) 사용한다. DataLoader의 'sampler'인자에 전달된다.

# {optimizer}.zero_grad()
[ 블로그 설명 ](https://algopoolja.tistory.com/55)
- gradient를 0으로 만들어 주는 코드
- Pytorch에서는 gradients값들을 추후에 backward를 해줄때 계속 더해주기 때문
- back propagation을 하기 전에는 항상 gradient를 0으로 만들어 줘야 한다.

# {model}.zero_grad()
[블로그 설명 ](https://minsuksung-ai.tistory.com/24)
- model의 weight를 전부 0으로 만들어 준다.

**Comment by doublejy715** [comment]
# Swaping subprocessing
[홈페이지](https://pytorch.org/docs/stable/multiprocessing.html)

multi gpu 에서 학습시키기 위한 코드.
function을 여러 gpu에 올려서 학습시키기 위해서 사용한다.

torch.multiprocessing.spawn

- fn : spawned process할 function을 의미. 
fn(i,*args)
i : process index
args : 넣어줄 매개변수
- args : fn에 넣어줄 매개변수. tuple로 묶어서 넣어준다.
- nprocs : 몇개의 sub process를 만들 것인지(보통 gpus의 개수로 설정)
- join : Perform a blocking join on all processes.(blocking join 이란?)

# dist.init_process_group

[공식 홈페이지 튜토리얼](https://tutorials.pytorch.kr/intermediate/dist_tuto.html)

# torch.utils.data.distributed.DistributedSampler
[ torch 공식 설명 ](https://pytorch.org/docs/stable/data.html)
[ 부가 설명 ](https://subinium.github.io/pytorch-dataloader/)
[ 부가 설명 2 ](https://medium.com/daangn/pytorch-multi-gpu-%ED%95%99%EC%8A%B5-%EC%A0%9C%EB%8C%80%EB%A1%9C-%ED%95%98%EA%B8%B0-27270617936b)

병렬 학습할때(분산 처리시) 사용한다. DataLoader의 'sampler'인자에 전달된다.

# {optimizer}.zero_grad()
[ 블로그 설명 ](https://algopoolja.tistory.com/55)
- gradient를 0으로 만들어 주는 코드
- Pytorch에서는 gradients값들을 추후에 backward를 해줄때 계속 더해주기 때문
- back propagation을 하기 전에는 항상 gradient를 0으로 만들어 줘야 한다.

# {model}.zero_grad()
[블로그 설명 ](https://minsuksung-ai.tistory.com/24)
- model의 weight를 전부 0으로 만들어 준다.

**Comment by doublejy715** [comment]
# torch.nn.parallel
## torch.nn.parallel.DistributedDataParallel
[ Official Link ](https://tutorials.pytorch.kr/intermediate/ddp_tutorial.html)
### 선언 
from torch.nn.parallel import DistributedDataParallel as DDP

ddp_model = DDP(model, device_ids)

#### DDP
[ Official Link ](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel)
```
DDP(module, device_ids=None, output_device=None, find_unused_parameters=False)
```

- model : 모델이나 Object를 넣어줘도 된다. 
- output_device = [{device_num}]
- find_unused_parameters = bool
반복을 할때 사용하지 않는 매개변수가 있는지를 찾아낸다.
model backward pass에 연관 되지 않는 parameter 들을 mark해서 DDP가 해당 파라미터들의 gradient들을 영원히 기다리는 것을 방지 한다. 
[Blog Link 1](https://study-grow.tistory.com/entry/pytorch-%EC%97%90%EB%9F%AC-DistributedDataParallel-%EC%97%90%EB%9F%AC)
[Blog Link 2](https://pajamacoder.tistory.com/14)

**Comment by doublejy715** [comment]
# Swaping subprocessing
[홈페이지](https://pytorch.org/docs/stable/multiprocessing.html)

multi gpu 에서 학습시키기 위한 코드.
function을 여러 gpu에 올려서 학습시키기 위해서 사용한다.

torch.multiprocessing.spawn

- fn : spawned process할 function을 의미. 
fn(i,*args)
i : process index
args : 넣어줄 매개변수
- args : fn에 넣어줄 매개변수. tuple로 묶어서 넣어준다.
- nprocs : 몇개의 sub process를 만들 것인지(보통 gpus의 개수로 설정)
- join : Perform a blocking join on all processes.(blocking join 이란?)

# dist.init_process_group

[공식 홈페이지 튜토리얼](https://tutorials.pytorch.kr/intermediate/dist_tuto.html)

# torch.utils.data.distributed.DistributedSampler
[ torch 공식 설명 ](https://pytorch.org/docs/stable/data.html)
[ 부가 설명 ](https://subinium.github.io/pytorch-dataloader/)
[ 부가 설명 2 ](https://medium.com/daangn/pytorch-multi-gpu-%ED%95%99%EC%8A%B5-%EC%A0%9C%EB%8C%80%EB%A1%9C-%ED%95%98%EA%B8%B0-27270617936b)

병렬 학습할때(분산 처리시) 사용한다. DataLoader의 'sampler'인자에 전달된다.

# {optimizer}.zero_grad()
[ 블로그 설명 ](https://algopoolja.tistory.com/55)
- gradient를 0으로 만들어 주는 코드
- Pytorch에서는 gradients값들을 추후에 backward를 해줄때 계속 더해주기 때문
- back propagation을 하기 전에는 항상 gradient를 0으로 만들어 줘야 한다.

# {model}.zero_grad()
[블로그 설명 ](https://minsuksung-ai.tistory.com/24)
- model의 weight를 전부 0으로 만들어 준다.

**Comment by doublejy715** [comment]
# Data Load
## DataLoader
- batch 기반의 딥러닝모델 학습을 위해서 mini batch를 만들어 주는 역할을 한다.
- dataloader를 통해 dataset의 전체 데이터가 batch size로 slice된다.
- 서버에서 돌릴 때는 num_worker를 조절해서 load를 올릴 수 있다.
(PC에서는 default로 성절해야 오류가 안난다.)

### Code
```
from torch.utils.data import Dataset

    val_loader = data.DataLoader(
        val_dataset,
        batch_size=args.batch_per_gpu,
        drop_last=True,
        sampler=val_sampler,
        num_workers=args.num_workers,
        pin_memory=True,
    )
```
#### Parameter 
- dataset
- drop_last : batch_size로 Dataset을 나눌 때 나머지가 있으면 자동으로 버려준다.

## torch.utils.data.distributed.DistributedSampler
- 분산처리를 위해서 사용하는 코드.
- torch.nn.parallel.DistributedDataParallel 과 함께 사용한다.

### Sampler
dataset은 index로 data를 가져오도록 설계되었기 때문에 shuffle을 하기 위해서 index를 적절히 섞어 주면 된다. 그것이 sampler이다.

### Process

전체 데이터를 GPU의 개수로 나눈다. 나눠진 부분 데이터에서만 DataLoader에서 데이터를 샘플링

1. 부분 데이터를 만들기 위해 전체 데이터셋 인덱스 리스트를 무작위로 섞는다.
2. 인덱스 리스트를 GPU개수만큼 쪼개서 각 GPU sampler에 할당한다.
3. epoch 마다 새롭게 GPU smapler에 무작위로 할당된다.

### Code
```
    val_sampler = torch.utils.data.distributed.DistributedSampler(
        dataset, num_replicas=world_size, rank=rank, shuffle=True
    )
```

#### instance
- dataset : dataset 묶음. 보통 torch.utils.data.Dataset을 사용한다.
- num_replicas : ngpus
- rank : index gpu

[ torch 공식 설명 ](https://pytorch.org/docs/stable/data.html)
[ 부가 설명 ](https://subinium.github.io/pytorch-dataloader/)
[ 부가 설명 2 ](https://medium.com/daangn/pytorch-multi-gpu-%ED%95%99%EC%8A%B5-%EC%A0%9C%EB%8C%80%EB%A1%9C-%ED%95%98%EA%B8%B0-27270617936b)
